src/microbe_stage/MembraneWaterRipple.cs

```diff
@@ -0,0 +1,666 @@
+        UpdateDetailLevelSubdivisions();
+
+        // Reset state tracking
+        currentPositionIndex = 0;
+        isPositionHistoryFull = false;
+
+        // Start informing state
+        isForming = true;
+        formingTime = 0.0f;
+        currentAlpha = 0.0f;
+        targetAlpha = 0.0f;
+        stillnessValue = 1.0f;
+        targetStillness = 1.0f;
+    }
+
+    private float CalculateTimeScale()
+    {
+        float timeScale = 1.0f;
+
+        if (currentSpeed > 0.1f)
+        {
+            // Scale based on distance from camera
+            if (isCameraPositionValid)
+            {
+                if (lastCameraDistance > 120.0f)
+                {
+                    timeScale = 1.0f + currentSpeed * 0.05f;
+                }
+                else if (lastCameraDistance > 80.0f)
+                {
+                    timeScale = 1.0f + currentSpeed * 0.08f;
+                }
+                else if (lastCameraDistance > 40.0f)
+                {
+                    timeScale = 1.0f + currentSpeed * 0.1f;
+                }
+                else
+                {
+                    timeScale = 1.0f + currentSpeed * 0.12f;
+                }
+            }
+        }
+
+        return Mathf.Lerp(timeScale, 0.5f, stillnessFactor);
+    }
+
+    /// <summary>
+    ///   Updates the ripple effect's alpha and stillness
+    /// </summary>
+    private void UpdateRippleEffect(float delta)
+    {
+        // Handle forming delay first
+        if (isForming)
+        {
+            formingTime += delta;
+            if (formingTime >= RippleFormationDelay)
+            {
+                isForming = false;
+                targetAlpha = fullAlpha;
+                targetStillness = 0.0f;
+            }
+
+            return;
+        }
+
+        targetAlpha = wasMovingLastFrame ? fullAlpha : minAlpha;
+        targetStillness = wasMovingLastFrame ? 0.0f : 1.0f;
+
+        // When newly not moving apply delay before fading out
+        if (!wasMovingLastFrame && stillnessTimer > StillnessFadeDelay)
+        {
+            targetAlpha = minAlpha;
+            targetStillness = 1.0f;
+        }
+
+        // Simple lerp with delta for transitions
+        currentAlpha = Mathf.Lerp(currentAlpha, targetAlpha,
+            delta * (currentAlpha < targetAlpha ? fadeInSpeed : fadeOutSpeed));
+        stillnessValue = Mathf.Lerp(stillnessValue, targetStillness, delta * 3.0f);
+        waterMaterial.SetShaderParameter(waterColorParam, new Color(0, 0, 0, currentAlpha));
+        waterMaterial.SetShaderParameter(stillnessFactorParam, stillnessValue);
+    }
+
+    /// <summary>
+    ///   Check if the effect is currently visible
+    /// </summary>
+    private bool IsVisible()
+    {
+        if (positionTarget == null || currentCamera == null || !positionTarget.IsInsideTree())
+            return false;
+
+        return GetViewport().GetVisibleRect().HasPoint(currentCamera.UnprojectPosition(positionTarget.GlobalPosition));
+    }
```

Comment:
Is updating the effect on the C# side already very expensive or why does this basically get around the normal frustrum culling that should already be in the engine?